bootstrap failed to build on Centos (devtoolset-7, 8)

**Describe the bug**
after commit :
e62d13612 ("[vcpkg] Add Supports: field. Use contents of triplets instead of names for dependency resolution. (#8601)", 2020-02-03)

**Environment**
- OS: Centos 7
- Compiler: Gcc 7, gcc 8 (devtoolset-7, devtoolset-8)

**To Reproduce**
Steps to reproduce the behavior:
1. ./bootstrap-vcpkg.sh
2. See error below, example is for devtoolset-8 (gcc8.3.1), same happens with devtoolset-7(gcc7xx). Interestingly, it does work on ubuntu1804 with the builtin gcc 8.3.0

**Expected behavior**
should build vcpkg
**Failure logs** 
```
[c7-local vcpkg]$ ./bootstrap-vcpkg.sh
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is GNU 8.3.1
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/rh/devtoolset-8/root/usr/bin/g++
-- Check for working CXX compiler: /opt/rh/devtoolset-8/root/usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Performing Test LINKS_TO_STDCXX_FS
-- Performing Test LINKS_TO_STDCXX_FS - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mitra/vcpkg/toolsrc/build.rel
[0/2] Re-checking globbed directories...
[14/69] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o
FAILED: CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o
/opt/rh/devtoolset-8/root/usr/bin/g++  -DDISABLE_METRICS=0 -I../include -O3 -DNDEBUG   -std=c++1z -pthread -MD -MT CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o -MF CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o.d -o CMakeFiles/vcpkglib.dir/src/vcpkg/binaryparagraph.cpp.o -c ../src/vcpkg/binaryparagraph.cpp
../src/vcpkg/binaryparagraph.cpp:31:5: note: ‘vcpkg::BinaryParagraph::BinaryParagraph()’ is implicitly deleted because the default definition would be ill-formed:
     BinaryParagraph::BinaryParagraph() = default;
     ^~~~~~~~~~~~~~~
../src/vcpkg/binaryparagraph.cpp:31:5: error: use of deleted function ‘vcpkg::PackageSpec::PackageSpec()’
In file included from ../include/vcpkg/binaryparagraph.h:3,
                 from ../src/vcpkg/binaryparagraph.cpp:7:
../include/vcpkg/packagespec.h:26:9: note: ‘vcpkg::PackageSpec::PackageSpec() noexcept’ is implicitly deleted because its exception-specification does not match the implicit exception-specification ‘’
         PackageSpec() noexcept = default;
         ^~~~~~~~~~~
../src/vcpkg/binaryparagraph.cpp: In constructor ‘vcpkg::BinaryParagraph::BinaryParagraph(vcpkg::Parse::RawParagraph)’:
../src/vcpkg/binaryparagraph.cpp:33:64: error: use of deleted function ‘vcpkg::PackageSpec::PackageSpec()’
     BinaryParagraph::BinaryParagraph(Parse::RawParagraph fields)
                                                                ^
../src/vcpkg/binaryparagraph.cpp: In constructor ‘vcpkg::BinaryParagraph::BinaryParagraph(const vcpkg::SourceParagraph&, const vcpkg::Triplet&, const string&, const std::vector<vcpkg::FeatureSpec>&)’:
../src/vcpkg/binaryparagraph.cpp:87:25: error: use of deleted function ‘vcpkg::PackageSpec::PackageSpec()’
         , type(spgh.type)
                         ^
../src/vcpkg/binaryparagraph.cpp: In constructor ‘vcpkg::BinaryParagraph::BinaryParagraph(const vcpkg::SourceParagraph&, const vcpkg::FeatureParagraph&, const vcpkg::Triplet&, const std::vector<vcpkg::FeatureSpec>&)’:
../src/vcpkg/binaryparagraph.cpp:98:101: error: use of deleted function ‘vcpkg::PackageSpec::PackageSpec()’
         : version(), description(fpgh.description), maintainer(), feature(fpgh.name), type(spgh.type)
                                                                                                     ^
[19/69] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o
ninja: build stopped: subcommand failed.
```

I will add that the workaround posted by @MarcKLohr still works as of the latest commit of Master, #11559  (tested on CentOS 7.6 (x64) with G++ 7.3 and 8.3 as provided in devtoolset-7 and devtoolset-8).

After I finished that, I did encounter 1 other error that *turned out to be unrelated.*  Due to a lack of adequate RAM (a total of 1GB) on the server where VCPKG was being compiled, the bootstrap script would fail at step 22:

```
[22/75] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o

FAILED: CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o
/opt/rh/devtoolset-8/root/usr/bin/g++  -DVCPKG_DISABLE_METRICS=0 -DVCPKG_USE_STD_FILESYSTEM=1 -I../include -O3 -DNDEBUG   -pthread -std=c++17 -MD -MT CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o -MF CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o.d -o CMakeFiles/vcpkglib.dir/src/vcpkg/build.cpp.o -c ../src/vcpkg/build.cpp
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.

[23/75] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/cmakevars.cpp.o
ninja: build stopped: subcommand failed.
```

I resolved this by creating a swap file:
```shell
sudo dd if=/dev/zero of=/swapfile bs=1024 count=1048576
sudo chmod 600 /swapfile
sudo mkswap /swapfile
sudo swapon /swapfile
```
With the extra "memory", the compiler was able to compile VCC correctly (`bootstrap-vcpkg.sh`).  After that, I removed the swapfile with: 
```
sudo swapoff /swapfile
sudo rm /swapfile
```

Hopefully this will help anyone having the same headaches as I did.

--- a/toolsrc/include/vcpkg/packagespec.h
+++ b/toolsrc/include/vcpkg/packagespec.h
@@ -20,8 +20,8 @@ namespace vcpkg
     ///
     struct PackageSpec
     {
-        PackageSpec() noexcept = default;
-        PackageSpec(std::string name, Triplet triplet) : m_name(std::move(name)), m_triplet(triplet) { }
+        PackageSpec() = default;
+        PackageSpec(std::string name, Triplet triplet) : m_name(std::move(name)), m_triplet(triplet) {}
 
         static std::vector<PackageSpec> to_package_specs(const std::vector<std::string>& ports, Triplet triplet);
 
@@ -54,7 +54,7 @@ namespace vcpkg
     ///
     struct FeatureSpec
     {
-        FeatureSpec(const PackageSpec& spec, const std::string& feature) : m_spec(spec), m_feature(feature) { }
+        FeatureSpec(const PackageSpec& spec, const std::string& feature) : m_spec(spec), m_feature(feature) {}
 
         const std::string& name() const { return m_spec.name(); }
         const std::string& feature() const { return m_feature; }
@@ -97,7 +97,7 @@ namespace vcpkg
         PackageSpec package_spec;
         std::vector<std::string> features;
 
-        FullPackageSpec() noexcept = default;
+        FullPackageSpec() = default;
         explicit FullPackageSpec(PackageSpec spec, std::vector<std::string> features = {})
             : package_spec(std::move(spec)), features(std::move(features))
         {

